### PR TITLE
[HLSL] Cleanup TargetInfo handling

### PR DESCRIPTION
**Pull request by llvm-beanz**
We had some odd places where we set target behaviors. We were setting the long size in target-specific code, but it should be language-based. We were not setting the Half float type semantics correctly, and instead were overriding the query in the AST context.

This change it moves existing code to the right places in the Target so that as we continue working on target and language feature they are controlled in the right places.

It also fixes a bug where the size of `half` was computed incorrectly when native half types are not supported.

### CODE REVIEW

File: clang/lib/Basic/TargetInfo.cpp [hunk 1]
```diff
@@ -406,6 +406,16 @@ void TargetInfo::adjust(DiagnosticsEngine &Diags, LangOptions &Opts) {
     LongDoubleAlign = 64;
   }
 
+  // HLSL explicitly defines the sizes and formats of some data types, and we
+  // need to conform to those regardless of what architecture you are targeting.
+  if (Opts.HLSL) {
+    LongWidth = LongAlign = 64;
+    if (!Opts.NativeHalfType) {
+      HalfFormat = &llvm::APFloat::IEEEsingle();
+      HalfWidth = HalfAlign = 32;
```
Comment (llvm-beanz) [inline clang/lib/Basic/TargetInfo.cpp:415]: Just as a note, we don't actually use this value anywhere in the HLSL code generation paths (yet), which is why it wasn't causing problems that we overrode the format returned by the AST context, but didn't override these values.

**Review by damyanp — APPROVED** [review]

**Review by bogner — APPROVED** [review]

**Review by farzonl — APPROVED** [review]

File: clang/lib/Basic/TargetInfo.cpp (hunk 1 above)
Comment (farzonl) [reply clang/lib/Basic/TargetInfo.cpp:415]: Is there a test case we can add for this?

File: clang/lib/AST/ASTContext.cpp [hunk 2]
```diff
@@ -1612,15 +1612,7 @@ const llvm::fltSemantics &ASTContext::getFloatTypeSemantics(QualType T) const {
...
-    // float.
-    if (getLangOpts().HLSL)
-      if (getLangOpts().NativeHalfType)
-        return Target->getHalfFormat();
-      else
-        return Target->getFloatFormat();
-    else
-      return Target->getHalfFormat();
+    return Target->getHalfFormat();
   case BuiltinType::Float:      return Target->getFloatFormat();
```
Comment (coopp) [inline clang/lib/AST/ASTContext.cpp:1616]: Existing (not in your new changes):  This switch statement is strangely formatted.  I seem some cases with the return on the same line as the case: and others on the next line.  Maybe this is how clang-format handles this?

Example:

case BuiltinType::Double:     return Target->getDoubleFormat();
  case BuiltinType::Ibm128:
    return Target->getIbm128Format();

**Review by coopp — APPROVED** [review]
Just a small nit from me about switch( ) statement formatting in existing code.  Not the new code.  Looks good to me.

File: clang/lib/AST/ASTContext.cpp (hunk 2 above)
Comment (llvm-beanz) [reply clang/lib/AST/ASTContext.cpp:1616]: Yea, this code would completely change with clang-format, but the bot only cares about the lines I changed. I tried not to include larger whitespace changes.

File: clang/lib/Basic/TargetInfo.cpp (hunk 1 above)
Comment (llvm-beanz) [reply clang/lib/Basic/TargetInfo.cpp:415]: Yes, and I totally glossed over that this fixes a bug. Without this line `sizeof(half)` returns the wrong value.

**Pull request merged**